be/src/storage/rowset_column_update_state.cpp

```diff
@@ -0,0 +1,524 @@
+        _update_chunk_cache[idx] = std::move(ChunkHelper::new_chunk(update_schema, DEFAULT_CONTAINER_SIZE));
+        chunk_shared_ptr = ChunkHelper::new_chunk(update_schema, DEFAULT_CONTAINER_SIZE);
+        ASSIGN_OR_RETURN(itrs, rowset->get_update_file_iterators(update_schema, &stats));
+    } else {
+        _update_chunk_cache[idx] = std::move(ChunkHelper::new_chunk(update_schema, 0));
+        chunk_shared_ptr = ChunkHelper::new_chunk(pkey_schema, DEFAULT_CONTAINER_SIZE);
+        ASSIGN_OR_RETURN(itrs, rowset->get_update_file_iterators(pkey_schema, &stats));
+    }
+
+    CHECK(itrs.size() == rowset->num_update_files()) << "itrs.size != num_update_files";
+    auto chunk = chunk_shared_ptr.get();
+    auto& dest = _upserts[idx];
+    auto& dest_chunk = _update_chunk_cache[idx];
+    auto col = pk_column->clone();
+    auto itr = itrs[idx].get();
+    if (itr != nullptr) {
+        col->reserve(DEFAULT_CONTAINER_SIZE);
+        while (true) {
+            chunk->reset();
+            auto st = itr->get_next(chunk);
+            if (st.is_end_of_file()) {
+                break;
+            } else if (!st.ok()) {
+                return st;
+            } else {
+                PrimaryKeyEncoder::encode(pkey_schema, *chunk, 0, chunk->num_rows(), col.get());
+                if (_enable_preload_column_mode_update_data) {
+                    _memory_usage += chunk->memory_usage();
+                    dest_chunk->append(*chunk);
+                }
+            }
+        }
+    }
+    for (const auto& itr : itrs) {
+        itr->close();
+    }
+    dest = std::move(col);
+    // This is a little bit trick. If pk column is a binary column, we will call function `raw_data()` in the following
+    // And the function `raw_data()` will build slice of pk column which will increase the memory usage of pk column
+    // So we try build slice in advance in here to make sure the correctness of memory statistics
+    dest->raw_data();
+    _memory_usage += dest != nullptr ? dest->memory_usage() : 0;
+
+    return Status::OK();
+}
+
+Status RowsetColumnUpdateState::_do_load(Tablet* tablet, Rowset* rowset) {
+    auto span = Tracer::Instance().start_trace_txn_tablet("rowset_column_update_state_load", rowset->txn_id(),
+                                                          tablet->tablet_id());
+
+    for (size_t i = 0; i < rowset->num_update_files(); i++) {
+        RETURN_IF_ERROR(_load_upserts(rowset, i));
+    }
+
+    for (size_t i = 0; i < rowset->num_update_files(); i++) {
+        RETURN_IF_ERROR(_prepare_partial_update_states(tablet, rowset, i, true));
+    }
+
+    return Status::OK();
+}
+
+Status RowsetColumnUpdateState::_prepare_partial_update_states(Tablet* tablet, Rowset* rowset, uint32_t idx,
+                                                               bool need_lock) {
+    if (_partial_update_states.size() == 0) {
+        _partial_update_states.resize(rowset->num_update_files());
```

Comment:
Why not resize it all the time?
If _partial_updates_states is not 0 and not equals with rowset->num_update_files, that will cause issuse.

Reply:
Because `num_update_files()` is const, resize once is enough. if `partial_update_states.size() != 0`, DCHECK it with `num_update_files()`